api/include/opentelemetry/trace/semantic_conventions.h

```diff
@@ -0,0 +1,190 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/version.h"
+
+#define OTEL_TRACE_GENERATE_SEMCONV_METHOD(attribute_name, attribute_value) \
+  static const nostd::string_view Get##attribute_name() noexcept            \
+  {                                                                         \
+    static const nostd::string_view attribute_name = attribute_value;       \
+    return attribute_name;                                                  \
+  }
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace trace
+{
+
+/**
+ * Stores the Constants for semantic kAttribute names outlined by the OpenTelemetry specifications.
+ * <see
+ * href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md"/>.
+ */
+class SemanticConventions final
+{
+public:
+  // The set of constants matches the specification as of this commit.
+  // https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md
+
+  // Genetal network connection attributes
+  OTEL_TRACE_GENERATE_SEMCONV_METHOD(AttributeNetTransport, "net.transport")
+  OTEL_TRACE_GENERATE_SEMCONV_METHOD(AttributeNetPeerIp, "net.peer.ip")
```

Comment:
Could we generate these definitions from the YAML definition of semantic conventions such as https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/http.yaml?

Reply:
...and could we generate them, and make them `constexpr` string constants? These would be encouraged by Google C++ coding style, and provide the most optimal compile-time optimization. `Global strings: if you require a global or static string constant, consider using a simple character array, or a char pointer to the first element of a string literal. String literals have static storage duration already and are usually sufficient.` .. constexpr string literal or string array should work, with all of these constants having unique prefix. Rather than grouping them in class - we group them in namespace. There is also a compile time flag that allows string pooling, we use that to avoid bloating in case if the same header gets included in different compilation units. For Visual Studio compiler that is `/GF (Eliminate Duplicate Strings)`. And for gcc / clang I think it's on by default?

Reply:
> Could we generate these definitions from the YAML definition of semantic conventions such as

I did think of writing a `script` to parse the yaml file and generate C++ code containing these `consts` from that and add it as part of CI. But this data is not going to change on daily basis, and instead, it would be fine to modify our code a couple of times before GA to be consistent. 
For otel-python and otel-go, they do have a script to generate code for these constants, along with other semantic data ( eg creating enums for allowable values for attributes, defining constraints of the list of attributes to be used etc) which we don't need to do. The current approach in this PR is as done by `otel-dotnet`.

Reply:
@ThomsonTan  - Have created issue for automatic generation of these definitions : https://github.com/open-telemetry/opentelemetry-cpp/issues/873
Once we finalize the approach to use, first ( and probably initial few ) definitions can be committed manually while someone works on script to generate code as per the finalised approach.